java-frontend/src/main/java/org/sonar/java/resolve/Resolve.java

```diff
@@ -478,41 +429,39 @@ private Resolution findMethod(Env env, JavaType callSite, JavaType site, String
     return bestSoFar;
   }
 
-  private JavaType handleTypeArguments(List<JavaType> typeParams, JavaType type, JavaSymbol.MethodJavaSymbol methodSymbol) {
-    if (!typeParams.isEmpty() && methodSymbol.typeVariableTypes.size() == typeParams.size()) {
-      TypeSubstitution typeSubstitution = new TypeSubstitution();
-      int i = 0;
-      for (JavaType.TypeVariableJavaType typeVariableType : methodSymbol.typeVariableTypes) {
-        typeSubstitution.add(typeVariableType, typeParams.get(i));
-        i++;
-      }
-      return substituteTypeParameter(type, typeSubstitution);
-    }
-    return type;
+  private static boolean isConstructor(JavaSymbol symbol) {
+    return "<init>".equals(symbol.name);
   }
 
   /**
    * @param candidate    candidate
    * @param bestSoFar previously found best match
    */
-  private JavaSymbol selectBest(Env env, JavaType site, List<JavaType> argTypes, JavaSymbol candidate, JavaSymbol bestSoFar, boolean autoboxing) {
+  private JavaSymbol selectBest(Env env, JavaType site, List<JavaType> argTypes, List<JavaType> typeParams, JavaSymbol candidate, JavaSymbol bestSoFar, boolean autoboxing) {
```

Comment:
The type of this method should be Resolution so there won't be a need to recompute/infer type substitution for return type as it is the case now. 
This will also allow simplification of findmethod